ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/AuthHandlers.java

```diff
@@ -52,51 +57,73 @@ static void setupAuthHandlers(final Server server, final Router router,
 
     systemAuthenticationHandler.ifPresent(handler -> router.route().handler(handler));
 
-    if (jaasAuthHandler.isPresent() || authenticationPlugin.isPresent()) {
-      router.route().handler(AuthHandlers::pauseHandler);
+    if (jaas.isPresent() || authenticationPlugin.isPresent()) {
+      router.route().handler(rc -> selectHandler(rc, jaas.isPresent(), pluginHandler.isPresent()));
 
-      router.route().handler(rc -> wrappedAuthHandler(rc, jaasAuthHandler, pluginHandler));
+      // set up using JAAS
+      jaas.ifPresent(h -> router.route().handler(wrappedHandler(h, Provider.JAAS)));
+      router.route().handler(wrappedHandler(
+          new RoleBasedAuthZHandler(
+              server.getConfig().getList(KsqlRestConfig.AUTHENTICATION_ROLES_CONFIG)),
+          Provider.JAAS));
 
+      // set up using PLUGIN
+      pluginHandler.ifPresent(h -> router.route().handler(wrappedHandler(h, Provider.PLUGIN)));
+
+      router.route().handler(AuthHandlers::pauseHandler);
```

Comment:
I thought order was important here -- that the pauseHandler had to come before either the JAAS or plugin handler. Did this still work even after the order was flipped?

Reply:
ordering is important here, which is why I needed to move it around. The new default implementation of the `BasicAuthHandler.create` that we plug into now calls pause/resume for you (meaning that it was resumed before it got to the `KsqlAuthorizationProviderHandler`).

Reply:
But it's not just the `KsqlAuthorizationProviderHandler` that the context needs to be paused for, right? I think we also want it paused for the `pluginHandler` and the `RoleBasedAuthZHandler` too. 

From the docs, the request needs to be paused before any async calls are made: https://vertx.io/docs/vertx-web/java/#_request_body_handling

The `pluginHandler` is a user's custom auth handler implementation, so we can't control whether async calls are being made or not.

`RoleBasedAuthZHandler` calls out to `AuthorizationHandler.create(...)` which I believe makes it [here](
https://github.com/vert-x3/vertx-web/blob/2f9905b36ff3cbc1f9cb6275bcc3f7a192370a0e/vertx-web/src/main/java/io/vertx/ext/web/handler/impl/AuthorizationHandlerImpl.java#L107), and this looks to be an async call.

Reply:
It might be safer to just pause for all handlers _except_ the one handler which has its own pause/resume logic built-in. That way we don't have to reason about which ones need to be paused for and which don't.

Reply:
makes sense, I'll make that change. For some reason I thought I followed that code path and it worked but now I can't do that again so either there were refactors I made that don't work or it never worked from the beginning.

Thanks for the review!